File: server/src/main/java/de/zalando/zally/apireview/ApiViolationsController.java
package de.zalando.zally.apireview;

import de.zalando.zally.dto.ApiDefinitionRequest;
import de.zalando.zally.dto.ApiDefinitionResponse;
import de.zalando.zally.dto.ViolationDTO;
import de.zalando.zally.dto.ViolationsCounter;
import de.zalando.zally.exception.MissingApiDefinitionException;
import de.zalando.zally.exception.UnaccessibleResourceUrlException;
import de.zalando.zally.rule.ApiValidator;
import de.zalando.zally.rule.Result;
import de.zalando.zally.rule.RulesPolicy;
import de.zalando.zally.rule.api.Severity;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.CrossOrigin;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestHeader;
import org.springframework.web.bind.annotation.ResponseBody;
import org.springframework.web.bind.annotation.RestController;

import java.util.Arrays;
import java.util.List;
import java.util.Map;

import static java.util.stream.Collectors.toList;
import static java.util.stream.Collectors.toMap;

@CrossOrigin
@RestController
public class ApiViolationsController {

    private final ApiValidator rulesValidator;
    private final ApiDefinitionReader apiDefinitionReader;
    private final ApiReviewRepository apiReviewRepository;
    private final ServerMessageService serverMessageService;
    private final RulesPolicy configPolicy;

    @Autowired
    public ApiViolationsController(ApiValidator rulesValidator,
                                   ApiDefinitionReader apiDefinitionReader,
                                   ApiReviewRepository apiReviewRepository,
                                   ServerMessageService serverMessageService,
                                   RulesPolicy configPolicy) {

        this.rulesValidator = rulesValidator;
        this.apiDefinitionReader = apiDefinitionReader;
        this.apiReviewRepository = apiReviewRepository;
        this.serverMessageService = serverMessageService;
        this.configPolicy = configPolicy;
    }

    @ResponseBody
    @PostMapping("/api-violations")
    public ApiDefinitionResponse validate(@RequestBody ApiDefinitionRequest request,
                                          @RequestHeader(value = "User-Agent", required = false) String userAgent) {
        String apiDefinition = retrieveApiDefinition(request);

        RulesPolicy requestPolicy = retrieveRulesPolicy(request);

        List<Result> violations = rulesValidator.validate(apiDefinition, requestPolicy);
        apiReviewRepository.save(new ApiReview(request, userAgent, apiDefinition, violations));

        return buildApiDefinitionResponse(violations, userAgent);
    }

    private RulesPolicy retrieveRulesPolicy(ApiDefinitionRequest request) {
        final List<String> requestRules = request.getIgnoreRules();
        if (requestRules == null) {
            return configPolicy;
        } else {
            return configPolicy.withMoreIgnores(requestRules);
        }
    }

    private String retrieveApiDefinition(ApiDefinitionRequest request) {
        try {
            return apiDefinitionReader.read(request);
        } catch (MissingApiDefinitionException | UnaccessibleResourceUrlException e) {
            apiReviewRepository.save(new ApiReview(request));
            throw e;
        }
    }

    private ApiDefinitionResponse buildApiDefinitionResponse(List<Result> violations, String userAgent) {
        ApiDefinitionResponse response = new ApiDefinitionResponse();
        response.setMessage(serverMessageService.serverMessage(userAgent));
        response.setViolations(violations.stream().map(this::toDto).collect(toList()));
        response.setViolationsCount(buildViolationsCount(violations));
        return response;
    }

    private ViolationDTO toDto(Result violation) {
        return new ViolationDTO(
            violation.getRule().title(),
            violation.getDescription(),
            violation.getViolationType(),
            violation.getRuleSet().url(violation.getRule()).toString(),
            violation.getPaths()
        );
    }

    private Map<String, Integer> buildViolationsCount(List<Result> violations) {
        ViolationsCounter counter = new ViolationsCounter(violations);
        return Arrays.stream(Severity.values()).collect(toMap(
            violationType -> violationType.toString().toLowerCase(),
            counter::getCounter
        ));
    }
}

File: server/src/main/java/de/zalando/zally/apireview/ApiViolationsController.kt
package de.zalando.zally.apireview

import de.zalando.zally.dto.ApiDefinitionRequest
import de.zalando.zally.dto.ApiDefinitionResponse
import de.zalando.zally.dto.ViolationDTO
import de.zalando.zally.dto.ViolationsCounter
import de.zalando.zally.exception.InaccessibleResourceUrlException
import de.zalando.zally.exception.MissingApiDefinitionException
import de.zalando.zally.rule.ApiValidator
import de.zalando.zally.rule.Result
import de.zalando.zally.rule.RulesPolicy
import de.zalando.zally.rule.api.Severity
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.web.bind.annotation.CrossOrigin
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestHeader
import org.springframework.web.bind.annotation.ResponseBody
import org.springframework.web.bind.annotation.RestController

@CrossOrigin
@RestController
class ApiViolationsController @Autowired
constructor(
    private val rulesValidator: ApiValidator,
    private val apiDefinitionReader: ApiDefinitionReader,
    private val apiReviewRepository: ApiReviewRepository,
    private val serverMessageService: ServerMessageService,
    private val configPolicy: RulesPolicy
) {

    @ResponseBody
    @PostMapping("/api-violations")
    fun validate(
        @RequestBody(required = true) request: ApiDefinitionRequest,
        @RequestHeader(value = "User-Agent", required = false) userAgent: String?
    ): ApiDefinitionResponse {
        val apiDefinition = retrieveApiDefinition(request)

        val requestPolicy = retrieveRulesPolicy(request)

        val violations = rulesValidator.validate(apiDefinition!!, requestPolicy)
        apiReviewRepository.save(ApiReview(request, userAgent.orEmpty(), apiDefinition, violations))

        return buildApiDefinitionResponse(violations, userAgent)
    }

    private fun retrieveRulesPolicy(request: ApiDefinitionRequest): RulesPolicy = request.ignoreRules
        ?.let { configPolicy.withMoreIgnores(request.ignoreRules!!) }
        ?: configPolicy

    private fun retrieveApiDefinition(request: ApiDefinitionRequest): String? = try {
        apiDefinitionReader.read(request)
    } catch (e: MissingApiDefinitionException) {
        apiReviewRepository.save(ApiReview(request, "", ""))
        throw e
    } catch (e: InaccessibleResourceUrlException) {
        apiReviewRepository.save(ApiReview(request, "", ""))
        throw e
    }

    private fun buildApiDefinitionResponse(violations: List<Result>, userAgent: String?): ApiDefinitionResponse {
        val response = ApiDefinitionResponse()
        response.message = serverMessageService.serverMessage(userAgent)
        response.violations = violations.map { this.toDto(it) }
        response.violationsCount = buildViolationsCount(violations)
        return response
    }

    private fun toDto(violation: Result): ViolationDTO = ViolationDTO(
        violation.rule.title,
        violation.description,
        violation.violationType,
        violation.ruleSet.url(violation.rule).toString(),
        violation.paths,
        if (violation.pointer == null) null else violation.pointer.toString()
    )

    private fun buildViolationsCount(violations: List<Result>): Map<String, Int> {
        val counter = ViolationsCounter(violations)
        return Severity.values()
            .map { severity ->
                severity.name.toLowerCase() to counter[severity]
            }
            .toMap()
    }
}
